From 51b5f218b9a7b238b1d064197d6fc6b50b073d7b Mon Sep 17 00:00:00 2001
From: Aaron Birkland <aaron.birkland@solo.io>
Date: Tue, 10 May 2022 11:47:09 -0400
Subject: Only expand vhost domains for local services (#38713)

* Only expand vhost domains for local services

When building outbound sidecar config, generate an expanded set of
hostnames (e.g. fqdn + short hostnames) only for local kubernetes
services in a given cluster.

* use DNSDomain instead of isClusterLocal(svc)

* gofumpt and gocritic

* release notes

* Clarify comments

* fix release notes

* fix unintended commit

* fix release notes
---
 .../pkg/networking/core/v1alpha3/httproute.go | 21 ++++++++++++++++++-
 .../core/v1alpha3/httproute_test.go           | 15 +++++++++++++
 releasenotes/notes/38678.yaml                 |  8 +++++++
 3 files changed, 43 insertions(+), 1 deletion(-)
 create mode 100644 releasenotes/notes/38678.yaml

diff --git a/pilot/pkg/networking/core/v1alpha3/httproute.go b/pilot/pkg/networking/core/v1alpha3/httproute.go
index 59491196ac..8cfac83c1b 100644
--- a/pilot/pkg/networking/core/v1alpha3/httproute.go
+++ b/pilot/pkg/networking/core/v1alpha3/httproute.go
@@ -533,8 +533,18 @@ func generateVirtualHostDomains(service *model.Service, port int, node *model.Pr
 // - Given foo.local.campus.net on proxy domain "" or proxy domain example.com, this
 // function returns nil
 func GenerateAltVirtualHosts(hostname string, port int, proxyDomain string) []string {
+	// If the dns/proxy domain contains `.svc`, only services following the <ns>.svc.<suffix>
+	// naming convention and that share a suffix with the domain should be expanded.
 	if strings.Contains(proxyDomain, ".svc.") {
-		return generateAltVirtualHostsForKubernetesService(hostname, port, proxyDomain)
+
+		if strings.HasSuffix(hostname, removeSvcNamespace(proxyDomain)) {
+			return generateAltVirtualHostsForKubernetesService(hostname, port, proxyDomain)
+		}
+
+		// Hostname is not a kube service.  It is not safe to expand the
+		// hostname as non-fully-qualified names could conflict with expansion of other kube service
+		// hostnames
+		return nil
 	}
 
 	var vhosts []string
@@ -729,3 +739,12 @@ func buildCatchAllVirtualHost(node *model.Proxy) *route.VirtualHost {
 		IncludeRequestAttemptCount: true,
 	}
 }
+
+// Simply removes everything before .svc, if present
+func removeSvcNamespace(domain string) string {
+	if idx := strings.Index(domain, ".svc."); idx > 0 {
+		return domain[idx:]
+	}
+
+	return domain
+}
diff --git a/pilot/pkg/networking/core/v1alpha3/httproute_test.go b/pilot/pkg/networking/core/v1alpha3/httproute_test.go
index 9ccd8c8b48..b6d5f65b73 100644
--- a/pilot/pkg/networking/core/v1alpha3/httproute_test.go
+++ b/pilot/pkg/networking/core/v1alpha3/httproute_test.go
@@ -111,6 +111,21 @@ func TestGenerateVirtualHostDomains(t *testing.T) {
 				"echo.default:8123",
 			},
 		},
+		{
+			name: "non-k8s service",
+			service: &model.Service{
+				Hostname:     "foo.default.svc.bar.baz",
+				MeshExternal: false,
+			},
+			port: 8123,
+			node: &model.Proxy{
+				DNSDomain: "default.svc.cluster.local",
+			},
+			want: []string{
+				"foo.default.svc.bar.baz",
+				"foo.default.svc.bar.baz:8123",
+			},
+		},
 		{
 			name: "k8s service with default domain and different namespace",
 			service: &model.Service{
diff --git a/releasenotes/notes/38678.yaml b/releasenotes/notes/38678.yaml
new file mode 100644
index 0000000000..d19c0cd7ac
--- /dev/null
+++ b/releasenotes/notes/38678.yaml
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 38678
+releaseNotes:
+  - |
+    **Fixed** some ServiceEntry hostnames can cause non-deterministic Envoy routes.
-- 
2.35.3

